File: TreeAppCoreC++/TreeWriters.cpp
#include "pch.h"
#include "TreeWriters.h"


DatabaseTreeWriter::DatabaseTreeWriter(string databaseName)
	:_databaseName(databaseName)
{
}

DatabaseTreeWriter::~DatabaseTreeWriter()
{
}

	void DatabaseTreeWriter::WriteTree(TreeNode& rootNode)
	{
		sqlite3* conn = DtabaseUtil::openDatabaseConnection(_databaseName);
		CleanupTables(conn);
		WriteNode(rootNode, TreeNode::ROOT_NODE_ID, conn);
		DtabaseUtil::closeDatabseConnection(conn);
	}

	void DatabaseTreeWriter::WriteNode(TreeNode& node, int parentId, sqlite3* conn)
	{
		WriteNodeInfo(node, conn);
		if (parentId != TreeNode::ROOT_NODE_ID)
		{
			WriteNodesRelation(node.Id, parentId, conn);
		}

		for (auto child : node.Children)
		{
			WriteNode(*child, node.Id, conn);
		}
	}

	void DatabaseTreeWriter::WriteNodeInfo(TreeNode& node, sqlite3* conn)
	{
		string query = "insert into TreeNodes(Id, Description) values(" + to_string(node.Id) + ", '" + node.Description + "')";
		DtabaseUtil::ExecuteNonQuery(query, conn);
	}

	void DatabaseTreeWriter::WriteNodesRelation(int nodeId, int parentId, sqlite3* conn)
	{
		string query = "insert into TreeNodeRelations(Id, ParentId) values(" + to_string(nodeId) + ", " + to_string(parentId) + ")";
		DtabaseUtil::ExecuteNonQuery(query, conn);
	}

	void DatabaseTreeWriter::CleanupTables(sqlite3* conn)
	{
		string query = "delete from TreeNodes";
		DtabaseUtil::ExecuteNonQuery(query, conn);

		query = "delete from TreeNodeRelations";
		DtabaseUtil::ExecuteNonQuery(query, conn);
	}

File: TreeAppFull/TreeWriters.cs
using System.Data.SqlClient;
using System.Linq;

namespace TreeAppFull
{
    public interface ITreeWriter
    {
        void WriteTree(TreeNode rootNode);
    }

    public class DatabaseTreeWriter : ITreeWriter
    {
        private readonly string _connectionString;

        public DatabaseTreeWriter(string connectionString)
        {
            _connectionString = connectionString;
        }

        public void WriteTree(TreeNode rootNode)
        {
            CleanupTables();

            WriteNode(rootNode, null);
        }

        private void WriteNode(TreeNode node, int? parentId)
        {
            WriteNodeInfo(node);
            if (parentId != null)
            {
                WriteNodesRelation(node.Id, parentId.Value);
            }

            foreach (var child in node.Children ?? Enumerable.Empty<TreeNode>())
            {
                WriteNode(child, node.Id);
            }
        }

        private void WriteNodeInfo(TreeNode node)
        {
            var sqlConn = new SqlConnection(_connectionString);
            var sqlComm = new SqlCommand($"insert into TreeNodes (Id, Description) values (@id, @description)", sqlConn);
            sqlComm.Parameters.AddWithValue("@id", node.Id);
            sqlComm.Parameters.AddWithValue("@description", node.Description);
            sqlComm.ExecuteNonQuery();
        }

        private void WriteNodesRelation(int nodeId, int parentId)
        {
            var sqlConn = new SqlConnection(_connectionString);
            var sqlComm = new SqlCommand($"insert into TreeNodeRelations (Id, ParentId) values (@id, @parentId)", sqlConn);
            sqlComm.Parameters.AddWithValue("@id", nodeId);
            sqlComm.Parameters.AddWithValue("@parentId", parentId);
            sqlComm.ExecuteNonQuery();
        }

        private void CleanupTables()
        {
            var sqlConn = new SqlConnection(_connectionString);
            var sqlComm = new SqlCommand("", sqlConn);
            sqlComm.CommandText = "delete from TreeNodes";
            sqlComm.ExecuteNonQuery();

            sqlComm.CommandText = "delete from TreeNodeRelations";
            sqlComm.ExecuteNonQuery();
        }
    }
}
